### PR TITLE
Task ring based message delivery implementation

### PR DESCRIPTION
**Pull request by Asitha**
## Task queue based message delivery implementation

- Introduce fair work division for message delivery (storage queue wise)
- Ability to add a delay to idle tasks

JIRA: https://wso2.org/jira/browse/MB-1671

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/task/TaskManager.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+     * Removes the {@link Task} with the given task id from the {@link Task} ring
+     *
+     * @param id  ID of the{@link Task} to be removed
+     */
+    public void remove(String id) {
+        if (log.isDebugEnabled()) {
+            log.debug("Task remove request. Task id" + id);
+        }
+
+
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/task/TaskManager.java]: We can remove extra new lines here

File: modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/ThriftConnectionListener.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.andes.thrift;
+
+/**
+ * Thrift connection listener interface
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/ThriftConnectionListener.java]: Shall we add more meaningful name here like what this class does. This comment is basically the class name.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/MBThriftClient.java [hunk 3]
```diff
@@ -351,31 +382,31 @@ private static void reConnectToServer() throws TTransportException, ClusterOutag
         }
     }
 
+    private static void notifyConnection() {
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/MBThriftClient.java]: Method level comment is missing.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Received slot for storage queue " + storageQueueName + " is: " +
+                                      currentSlot.getStartMessageId() + " - " + currentSlot.getEndMessageId() +
+                                      "Thread Id:" + Thread.currentThread().getId());
+                }
+                List<DeliverableAndesMetadata> messagesRead = getMetaDataListBySlot(storageQueueName, currentSlot);
+
+                if (messagesRead != null && !messagesRead.isEmpty()) {
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java]: Shall we change the if condition to something like below. 
if ((messagesRead != null) && (!messagesRead.isEmpty()))

**Comment by ramith** [comment]
Attach a Jira?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorkerManager.java [hunk 5]
```diff
@@ -206,8 +155,51 @@ public SlotDeliveryWorker getSlotWorker(String queueName) {
...
      */
     public void dumpAllSlotInformationToFile(File fileToWrite) throws AndesException {
-        for (SlotDeliveryWorker slotDeliveryWorker : slotDeliveryWorkerMap.values()) {
-            slotDeliveryWorker.dumpAllSlotInformationToFile(fileToWrite);
-        }
+        // NOTE: Will be replaced with subscription store update PR
+    }
+
+    @Override
+    public void minimumNodeCountNotFulfilled(int currentNodeCount) {
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorkerManager.java]: add method comments?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/task/TaskManager.java (hunk 1 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/task/TaskManager.java]: fixed

File: modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/ThriftConnectionListener.java (hunk 2 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/ThriftConnectionListener.java]: comment changed

File: modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/MBThriftClient.java (hunk 3 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/MBThriftClient.java]: Comment added

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java (hunk 4 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java]: Used[ CollectionUtills#isEmpty(...) ](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/CollectionUtils.html#isEmpty(java.util.Collection))

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorkerManager.java (hunk 5 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorkerManager.java]: Comment added

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.andes.kernel.slot;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java]: does this need to be in this package?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorker.java [hunk 7]
```diff
@@ -588,18 +588,17 @@ private void flagNetworkoutage() {
...
-	 * network partition is healed. therefore removing the barrier (- allows the this thread to work) 
-	 */
-	@Override
-	public void minimumNodeCountFulfilled(int currentNodeCount) {
-		log.info(this.getId() + " network outage resolved therefore resuming work, current cluster size: " + currentNodeCount);
-		networkOutageDetected.set(false);
-
-	}
+    /**
+     * network partition is healed. therefore removing the barrier (- allows the this thread to work)
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorker.java]: out dated code isn't it? shall we remove?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java (hunk 6 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/MessageDeliveryTask.java]: Still this reads a slot and deliver messages . Can be  named as `SlotDeliveryTask`.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorker.java (hunk 7 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeliveryWorker.java]: Removed `SlotDeliveryWorker`

**Pull request merged**